[jit] Fuse tensor-scalar ops when scalar is constant

zou3519 opened the pull request:
This is on the way to resolving #9940.

Fixes #10501

This PR modifies graph fuser to fuse operations that have constant
scalar arguments. These constant scalar arguments are directly inlined
into the kernel body.

The context for this is that LSTM backward (in particular, sigmoid
backward) has many add(x, 1.) operations. This PR should be sufficient for
LSTM backward to get fused by the graph fuser.

cc @apaszke @zdevito

zou3519 committed:
Fuse tensor-scalar ops when scalar is constant

This is on the way to resolving #9940.

This PR modifies graph fuser to fuse operations that have constant
scalar arguments. These constant scalar arguments are directly inlined
into the kernel body.

The context for this is that LSTM backward (in particular, sigmoid
backward) has many add(x, 1.) operations. This PR should be sufficient for
LSTM backward to get fused by the graph fuser.

zou3519 committed:
Expect file updates

zou3519 commented on test/test_jit.py:
```diff
@@ -1969,6 +1969,7 @@ def checkScript(self, script, inputs, optimize=True, outputs=None, name='func',
         else:
             outputs_ge = ge(*inputs)
         self.assertEqual(outputs, outputs_ge)
+        return ge
```
Will fix

zou3519 committed:
Fix tests

zou3519 committed:
Fix expect files

zdevito reviewed (commented):
This looks good. I have some small comments. I am surprised that we didn't need to modify the fusion_compiler, did we already have code to emit constants?

zou3519 committed:
Addressed comments:

- Use WithInsertPoint instead of insertAfter
- Make the compatible devices logic more explicit by adding a Device
  struct and a DeviceType enum. The possible Devices are
  `Unknown | AnyDevice | CPU | CUDA i`

zou3519 commented:
@zdevito Yes there was already code to emit constants inlined into the body of the FusionGroup. In particular, the `alpha` argument in `add(Tensor, Tensor, Scalar alpha)` was inlined into the FusionGroup if it is constant. I took this logic and expanded it to work with all constant number arguments.

zou3519 commented on test/expect/TestScript.test_lstm_fusion_cuda-backward.expect:
```diff
@@ -18,70 +18,56 @@ graph(%0 : Float(3, 20!)
-  %37 : Float(80, 10) = aten::mm(%36, %x)
-  return (%37, %35, %33, %33);
+  %20 : Float(3, 80) = prim::FusionGroup_0[device=0](%ingate, %forgetgate, %cellgate, %outgate, %cx, %1, %18, %0)
+  %21 : Float(3, 80) = aten::mul(%20, %19)
```
I don't know what this mul(tensor, 1) is doing here, but I can run CSE after the graph fuser to get rid of it (probably). Will investigate in another task.

apaszke replied:
This diff was supposed to work with `x * 1` too right?

zou3519 replied:
Yes, but the reason why that one doesn't get fused into the fusion group is because the output is a `prim::FusedConcat` node. The map_size of the fusion group is `(3, 20) `which doesn't fit the map_size of aten::mul `(3, 80)`

apaszke commented on test/test_jit.py:
```diff
@@ -2288,6 +2288,30 @@ def func2():
+
+        def should_not_fuse(x, z):
+            y = x + int(z)
+            return x * y
```
Why shouldn't we fuse this one?

zou3519 replied:
z isn't constant. This PR only inlines numbers into the graph fuser if they're constant.

apaszke replied:
ohhh makes sense. Can you please just drop in a tiny comment explaining that?

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -125,7 +170,7 @@ struct GraphFuser {
       #endif
     }
-    return false;
+    return allow_constant_numbers && node->type()->isSubtypeOf(NumberType::get());
```
You didn't check that the number is in fact a constant. Even if it's not a bug, we should at least change the name of the argument to `allow_numbers`.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -141,33 +186,51 @@ struct GraphFuser {
+          v->node()->kind() != prim::Constant) {
+        return false;
+      }
+    }
```
I don't really like this function, because we end up with the hacky checks we had previously. If we want to handle tensor-scalar ops, this function should explicitly look for them using `->matches()`, and have a fallback that expect all-tensor inputs for all other pointwise ops. It would be much clearer too. You can leave the code in `hasSupportedType(Node*)` as it was, and put the `matches` checks before to shortcut the scalar cases. This way we wouldn't need "allow constant numbers" either - you could just slice out the tensor inputs.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -141,33 +186,51 @@ struct GraphFuser {
-               node->matches("aten::ge(Tensor self, Tensor other) -> Tensor") ||
-               node->matches("aten::eq(Tensor self, Tensor other) -> Tensor") ||
-               node->matches("aten::ne(Tensor self, Tensor other) -> Tensor")) {
+    if (node->matches("aten::lt(Tensor self, Tensor other) -> Tensor") ||
```
Actually, since we have some `matches` code here, you might want to put it in there instead of in `hasSupportedType`

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -249,18 +312,48 @@ struct GraphFuser {
+        producer_device.type() == DeviceType::CPU) {
+      if (!sharedFusionCompiler().canCompileOnCPU()) {
+        return false;
+      }
```
The inner conditional is just `return sharedFusionCompiler.canCompileOnCPU()`.

zou3519 replied:
But it's possible that consumer_device is DeviceType::CPU and producer_device is DeviceType::CUDA. I wanted to check if the fusion compiler can compile on CPU before going on to the other conditional statements later.

This logic is a little convoluted so I'll try to think of a way to clean it up.

apaszke replied:
I'm not sure if I follow. I think if you were to replace the 3 lines above with what I proposed, then it would end up being 100% equivalent. Isn't that the case?

zou3519 replied:
Consider the case where consumer_device.Type() == DeviceType::CPU and producer_device.type() == CUDA:

If I changed
```
if (consumer_device.type() == DeviceType::CPU ||
        producer_device.type() == DeviceType::CPU) {
      if (!sharedFusionCompiler().canCompileOnCPU()) {
        return false;
      }
}
```
to 
```
if (consumer_device.type() == DeviceType::CPU ||
        producer_device.type() == DeviceType::CPU) {
      return sharedFusionCompiler.canCompileOnCPU()
   }
}
```
and `sharedFusionCompiler.canCompileOnCPU()` is true, then `compatibleDevices(...)` would return true. However, consumer_device is different from producer_device, so we should not be returning true here

apaszke replied:
Oh god, of course. Never mind.

apaszke reviewed (commented):
Mostly LGTM, but I'd like to remove the hacky scalar checks that don't even try to see what overloads do we handle. Please use the `matching` syntax, or we'll end up with a lot of bugs again.

zou3519 committed:
Address comments:

- Use ->matches instead of hacky allowing-numbers-in-type-checks
- Rewrite bool compatibleDevices(Node * consumer, Value * producer) to
  be more readable.

zou3519 commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -299,7 +299,7 @@ RegisterOperators reg({
 
 #define DEFINE_ST_OP(aten_op, reverse_exp)                             \
-  Operator("aten::" #aten_op "(Scalar a, Tensor b) -> Tensor", [](Node* node) { \
+  Operator("aten::" #aten_op "(Scalar other, Tensor self) -> Tensor", [](Node* node) { \
```
This was a necessary change because `attr::a` and `attr::b` don't exist (register_prim_ops doesn't inject symbols into aten_interned_strings). I think this rename makes more sense because one can do `torch.add(other=1, self=tensor)` in python land.

zou3519 committed:
Fix mul/div

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -146,30 +191,65 @@ struct GraphFuser {
   }
 
-  bool haveSupportedType(at::ArrayRef<Value*> list) {
+  bool haveSupportedType(at::ArrayRef<Value*> list, bool allow_constant_numbers=false) {
```
Is this argument used?

zou3519 replied:
Forgot to remove the allow_constant_numbers. Will remove

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -146,30 +191,65 @@ struct GraphFuser {
+        result.push_back(v);
+      }
+    }
+    return result;
```
nit: I feel like we have a `filter` function in `torch/csrc/utils/funcional.h`

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -249,18 +329,53 @@ struct GraphFuser {
+               consumer_device.type() == DeviceType::AnyDevice) {
+      // XXX: This case means we're fusing operations on non-constant numbers.
+      // The graph fuser doesn't support this at the moment (#9940).
+      JIT_ASSERT(false);
```
Why do we have a hard failure in here instead of returning false?

apaszke reviewed (approved):
Looks great now! Some minor comments.

zou3519 committed:
Address comments
